File: gogi/chunks/fixed_size_chunking.go
package chunks

import (
	"strings"

	"gogi/gogi/documents"
)

// Fixed-size chunking implementation
type FixedSizeChunking struct {
	ChunkSize    int
	ChunkOverlap int
}

// Constructor
func NewFixedSizeChunking(
	chunkSize int,
	chunkOverlap int,
) *FixedSizeChunking {

	if chunkSize <= 0 {
		chunkSize = 512
	}

	if chunkOverlap < 0 {
		chunkOverlap = 50
	}

	return &FixedSizeChunking{
		ChunkSize:    chunkSize,
		ChunkOverlap: chunkOverlap,
	}
}

// Chunk implements ChunkingStrategy
func (f *FixedSizeChunking) Chunk(
	document documents.ExtractedDocument,
) []Chunk {

	text := document.Text()

	if strings.TrimSpace(text) == "" {
		return []Chunk{}
	}

	words := strings.Fields(text)

	var chunks []Chunk

	step := f.ChunkSize - f.ChunkOverlap

	if step < 1 {
		step = 1
	}

	i := 0

	for i < len(words) {

		end := i + f.ChunkSize

		if end > len(words) {
			end = len(words)
		}

		chunkWords := words[i:end]
		chunkText := strings.Join(chunkWords, " ")

		// approximate search offset
		searchStart := 0

		for _, w := range words[:i] {
			searchStart += len(w) + 1
		}

		startOffset := searchStart

		if len(chunkWords) > 0 {
			found := strings.Index(
				text[searchStart:],
				chunkWords[0],
			)

			if found >= 0 {
				startOffset = searchStart + found
			}
		}

		endOffset := startOffset + len(chunkText)

		chunks = append(chunks, Chunk{
			Text:        chunkText,
			StartOffset: startOffset,
			EndOffset:   endOffset,
		})

		i += step
	}

	return chunks
}

File: workers/worker_utils/chunking/fixed_size_chunking.py
from typing import List, override
from loguru import logger

from workers.worker_utils.chunking.text_chunk_model import TextChunk
from workers.worker_utils.documents.extracted_document import ExtractedDocument
from workers.worker_utils.chunking.chunking_strategy_base import ChunkingStrategy, ChunkingType

class FixedSizeChunking(ChunkingStrategy):
    """Word-count-based fixed-size chunking"""

    def __init__(self):
        super().__init__(name=ChunkingType.FIXED)

    @override
    def chunk(self, document: ExtractedDocument, *, chunk_size: int = 512, chunk_overlap: int = 50) -> List[TextChunk]:


        logger.debug(f"Chunking document with strategy: {self.name} with chunk_size={chunk_size}, chunk_overlap={chunk_overlap}")

        text = document.text
        if not text.strip():
            return []

        words = text.split()
        chunks: List[TextChunk] = []
        step = max(1, chunk_size - chunk_overlap)
        i = 0

        while i < len(words):
            chunk_words = words[i : i + chunk_size]
            chunk_text = " ".join(chunk_words)
            search_start = sum(len(w) + 1 for w in words[:i])
            start_offset = text.find(chunk_words[0], search_start) if chunk_words else 0
            end_offset = start_offset + len(chunk_text)
            chunks.append(TextChunk(text=chunk_text, start_offset=start_offset, end_offset=end_offset))
            i += step

        return chunks